### DISCUSSION TITLE
关于插件分类的一些想法

### DISCUSSION THREAD

**Discussion by synodriver** [nonebot/discussions#94] [想法]
似乎目前的插件商店对各种插件的分类还比较混乱，各种tag都有，于是，以下提出一种根据插件功能进行分类的想法

# 主要分为三类插件

- A类插件： 注册matcher，可以对特定的```event```做出响应的（简单来说就是会说话的那种插件）。目前商店的大部分是这种类型的

- B类插件： 使用```export```功能，在自己的命名空间注册一些对象，方便别的插件使用的，主要用于跨插件调用。典型的，有[定时器插件](https://github.com/nonebot/plugin-apscheduler)

- C类插件：在```serverapp```上做文章的，这种插件主要是利用了nonebot作为webserver的功能，以实现别的需求。典型的，有[文档插件](https://github.com/nonebot/nonebot2/tree/master/packages/nonebot-plugin-docs)

这种分类就能一目了然的知道插件的性质了。比如，B类插件绝对没法开箱即用，它们的功能是辅助别的插件。通常，B类插件要优先于A类插件加载，否则可能出问题。又比如，正向driver就不能使用文档插件，（因为正向是舔狗，根本没有serverapp对象）。![](https://user-images.githubusercontent.com/25792361/110955661-c93be700-8384-11eb-9897-b1a4377acab5.png)![](https://user-images.githubusercontent.com/25792361/110955666-ca6d1400-8384-11eb-84c2-a5282a2962aa.png)

当然，这种分类也不是绝对的，如果一个插件有多种功能，那就应该是混合型的插件，（比如AB类，自己能说话的同时向别的插件暴露了一些功能）

**Comment by synodriver** [comment]
现在b1更新的影响就非常清楚了：A类插件全灭

**Reply by Lancercmd** [reply]
这样说就让人感到非常有趣，框架的破坏性改动，受影响最大的是最上层的功能插件，反而不是利用框架内部方法的下层插件。
> 影响是恶劣的，也就是基本没有人会去跟进框架的更新，就像你我能不升 Win11 就不升一样。
> 明明可以做到兼容的地方为什么就改了呢，会给人一种不可思议的感觉。
>
> 而这是插件商店最见不得的情况，插件开发者倾向于选择一个相对稳定的版本来跟进。
> 如此一来，直到 NoneBot2 真正地发布 Released 以前，都不会有开发者关心插件商店。
> 哪怕增加更多更友好的 `tag` 框架说不定明天就会提交 breaking changes 想到这点的插件开发者的安全感便完全为负。
>
> 即使如此，依然感谢即使如此也对插件商店做出贡献的插件开发者们。

**Comment by synodriver** [comment]
或许可以给插件商店的打上类似的tag？

**Comment by Lancercmd** [comment]
取决于插件商店的受众，商店插件的绝大部分应该都归属于A类。
> 数量上差距悬殊的话那么这样基本上就没什么好分的了。

所述的B类、C类插件一般需配合插件的 README 食用，一般也能通过插件的显示名称或说明看出来。
> 看出其不是一个即插即用型的，或者说功能型的插件。

同时商店插件的 `tag` 是可以被作为关键字搜索的，所以像现在这样仅通过插件开发者自己对插件标注 `tag` 也还行。
> 不过若是从预设 `tag` 的角度出发，那届时可能还需要让 @yanyongyu 重新设计页面来区分三类插件了（

**Comment by mnixry** [comment]
我感觉这个想法其实非常不错... 但是主要是现在这个前端也就只有 @yanyongyu 一个人写

我觉得前端需要加的功能:

- 允许用户在添加tag的时候搜索已有tag, 如果已经有了就选上, 没有跳一个添加新tag的界面
- 添加新tag需要几个内容: tag的namespace, 本来名称, 主题色, 一个简单的解释 (可以限制字符数)

同时在发布以上功能的同时, 我们需要添加几个预定义的tag
比如上述的ABC类tag 还有对应adapter的tag. 对应协议端实现的tag可以让用户自己添加

**Reply by mnixry** [reply]
另外, 这三类其实可以总结成三种类型

- A类插件: `ability:bot-service` 发挥作为Bot的能力, 提供具体业务
- B类插件: `ability:infrastructure` 作为基建插件, 为其他插件提供能力
- C类插件: `ability:adapter-service` 基于adapter的额外能力提供业务

比如我的[`nonebot-plugin-gocqhttp`](https://github.com/mnixry/nonebot-plugin-gocqhttp)应该同时属于A和C类

**Reply by Lancercmd** [reply]
> 比如我的 `nonebot-adapter-gocqhttp` 应该同时属于A和C类

这样的话感觉只要向上取C类即可，插件按最高级（低级？）用途取分类

**Reply by mnixry** [reply]
> 这样的话感觉只要向上取C类即可，插件按最高级（低级？）用途取分类

没有, 它真的有个指令, 用来获取进程状态的, 所以说它有Bot的能力也无可厚非

**Reply by Lancercmd** [reply]
> 没有, 它真的有个指令, 用来获取进程状态的, 所以说它有Bot的能力也无可厚非

就是说只要插件具备更高一类的预期用途，那就可以直接按更高一类区分，而不需要标注复数分类。  
话说回来，这样的 `tag` 分类方法作为强制执行的话，插件发布前的审核需要跟进。因为这是个概念型的分类方式，开发者对插件进行错误的分类并不是没有可能的。

> A类插件: `ability:bot-service` 发挥作为Bot的能力, 提供具体业务
> B类插件: `ability:infrastructure` 作为基建插件, 为其他插件提供能力
> C类插件: `ability:adapter-service` 基于adapter的额外能力提供业务
>
> A类插件往往也属于C类插件，是基于 Adapter 提供具体业务的。
> 但是显然纯粹的功能型插件就是A类插件，这就存在插件开发者对插件错误分类的可能性。
> 而像 @synodriver 的分类形式，其实就是说如果（按现在商店插件的情况）类目之间插件的数量差距悬殊，那么分类的意义就仅仅是多了个 `tag` 了，对于前端的显示效果优化依然是个挑战。
> 当然我也知道 `tag` 的意义并不仅仅是对前端的。

**Reply by synodriver** [reply]
C类插件应该指的是在```adapter```注册的```endpoint```之外新开```endpoint```的，这点可能没讲清楚

**Comment by synodriver** [comment]
```python
import asyncio
from typing import cast
import traceback

import nonebot
from nonebot.adapters.onebot.v11 import Adapter
from fastapi import Query, FastAPI
from fastapi.responses import UJSONResponse, HTMLResponse


app = nonebot.get_asgi()
app = cast(FastAPI, app)
driver = nonebot.get_driver()


@app.get('/')
async def handle_get():
    return HTMLResponse(f"""<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Nonebot</title>
</head>
<body>
<h1 align="center"><strong>nonebot{nonebot.__version__} is runing!</strong></h1>
<p align="center">Vist <a href="https://v2.nonebot.dev">noneot document</a> for more information</p>
<hr/>
<p align="center">Yours, nonebot</p>
</body>
</html>""")


@app.get('/bots')
async def get_bots():
    ret = {"bots": []}
    [ret["bots"].append(qq) for qq in driver._clients]
    return ret


@app.get("/disconnect")
async def disconnect(bot: str = Query(...)) -> UJSONResponse:
    try:

        adapter = nonebot.get_driver()._adapters[Adapter.get_name()]
        adapter = cast(Adapter, adapter)
        ws = adapter.connections[bot]
        await ws.close()
        adapter.connections.pop(bot, None)
        adapter.bot_disconnect(nonebot.get_bot(bot))
        nonebot.logger.info(f"disconnect bot: {bot}")
    except Exception as e:
        return UJSONResponse({"msg": "no such bot", "exc": traceback.format_exc()}, 404)
    else:
        return UJSONResponse({"msg": "ok"}, 200)
```

一个简单的C类插件，可以查看状态之类的，释放连接

**Reply by synodriver** [reply]
比较典型的应用（
